data/src/main/java/com/linkedin/data/collections/CheckedMap.java

```diff
@@ -541,8 +556,60 @@ private void notifyChangeListenersOnClear(Set<? extends K> keys)
     void onUnderlyingMapChanged(K key, V value);
   }
 
+  private void purgeStaleChangeListeners()
+  {
+    if (_changeListenerReferenceQueue != null && _changeListenerReferenceQueue.poll() != null)
+    {
+      // Clear finalized weak references.
+      while (_changeListenerReferenceQueue.poll() != null)
+      {
+        // Do nothing, as we are just clearing the reference queue.
```

Comment:
Sorry, new to this, but wont this cause infinite looping?

Reply:
poll removes the head of the reference queue.

Reply:
It calls queue.poll(), which remove the head of the queue. Once the queue is empty, it will return null, then the loop ends.

Reply:
Yeah this seems incorrect and also inefficient? Wouldn't it be better to just go over the referencequeue pull elements out one by one and delete them inline rather than iterate through the entire list?